File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/componets/spendingsections/SpendingSectionsFragment.java
package ru.strcss.projects.moneycalc.moneycalcandroid.componets.spendingsections;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.android.support.DaggerFragment;
import moneycalcandroid.moneycalc.projects.strcss.ru.moneycalc.R;
import ru.strcss.projects.moneycalc.moneycalcandroid.componets.spendingsections.addeditspendingsection.AddEditSpendingSectionActivity;
import ru.strcss.projects.moneycalc.moneycalcandroid.storage.DataStorage;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SpendingSection;

import static ru.strcss.projects.moneycalc.moneycalcandroid.utils.view.UiUtils.showProgress;

public class SpendingSectionsFragment extends DaggerFragment implements SpendingSectionsContract.View {

    @Inject
    SpendingSectionsContract.Presenter presenter;

    @Inject
    DataStorage dataStorage;

    @Inject
    public SpendingSectionsFragment() {
    }

    // UI references
    private FloatingActionButton fabAddSpendingSection;
    private RecyclerView rvSpendingSections;
    private SpendingSectionsAdapter adapter;
    private List<SpendingSection> sectionList;
    private ProgressBar progressView;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.spendingsections_frag, container, false);

        rvSpendingSections = root.findViewById(R.id.rv_spendingsections);
        progressView = root.findViewById(R.id.spending_section_progress);

        sectionList = new ArrayList<>();
        adapter = new SpendingSectionsAdapter(getContext(), sectionList, presenter);

        RecyclerView.LayoutManager mLayoutManager = new GridLayoutManager(getContext(), 1);
        rvSpendingSections.setLayoutManager(mLayoutManager);
//        rvSpendingSections.addItemDecoration(new GridSpacingItemDecoration(2, dpToPx(10), true));
        rvSpendingSections.setItemAnimator(new DefaultItemAnimator());
        rvSpendingSections.setAdapter(adapter);

        presenter.requestSpendingSections();

        fabAddSpendingSection = root.findViewById(R.id.fab_spendingsecions_addspendingsection);
        fabAddSpendingSection.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(getContext(), AddEditSpendingSectionActivity.class);
                startActivityForResult(intent, 0);
            }
        });
        rvSpendingSections.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0)
                    fabAddSpendingSection.hide();
                else if (dy < 0)
                    fabAddSpendingSection.show();
            }
        });

        // show available data quickly
        if (dataStorage.getSpendingSections() != null) {
            showSpendingSections(dataStorage.getSpendingSections());
        }
        return root;
    }

    @Override
    public void showSpendingSections(List<SpendingSection> spendingSections) {
        adapter.updateList(spendingSections);
    }

    @Override
    public void showErrorMessage(String msg) {
        System.out.println("showErrorMessage! " + msg);
        Snackbar.make(getActivity().findViewById(android.R.id.content), msg, Snackbar.LENGTH_LONG).show();
    }

    @Override
    public void showSpinner() {

    }

    @Override
    public void hideSpinner() {
        int animTime = getResources().getInteger(android.R.integer.config_mediumAnimTime);
        showProgress(false, null, progressView, animTime);
    }

    @Override
    public void showUpdateSuccess() {
        Snackbar.make(getActivity().findViewById(android.R.id.content), R.string.spending_section_edit_success, Snackbar.LENGTH_LONG).show();
        // TODO: 05.06.2018 show mini spinner here
        presenter.requestSpendingSections();
    }

    @Override
    public void showDeleteSuccess() {
        Snackbar.make(getActivity().findViewById(android.R.id.content), R.string.spending_section_delete_success, Snackbar.LENGTH_LONG).show();
        // TODO: 05.06.2018 show mini spinner here
        presenter.requestSpendingSections();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.takeView(this);
//        presenter.requestSpendingSections();
//        System.out.println("SpendingSectionsFragment onResume");
    }

    @Override
    public void onDestroy() {
        presenter.dropView();
        super.onDestroy();
    }
}

File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/componets/spendingsections/SpendingSectionsFragment.kt
package ru.strcss.projects.moneycalc.moneycalcandroid.componets.spendingsections

import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import dagger.android.support.DaggerFragment
import moneycalcandroid.moneycalc.projects.strcss.ru.moneycalc.R
import ru.strcss.projects.moneycalc.moneycalcandroid.componets.spendingsections.addeditspendingsection.AddEditSpendingSectionActivity
import ru.strcss.projects.moneycalc.moneycalcandroid.storage.DataStorage
import ru.strcss.projects.moneycalc.moneycalcandroid.utils.view.UiUtils.showProgress
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SpendingSection
import java.util.*
import javax.inject.Inject

class SpendingSectionsFragment @Inject
constructor() : DaggerFragment(), SpendingSectionsContract.View {

    @Inject
    lateinit var presenter: SpendingSectionsContract.Presenter

    @Inject
    lateinit var dataStorage: DataStorage

    // UI references
    private var fabAddSpendingSection: FloatingActionButton? = null
    private var rvSpendingSections: RecyclerView? = null
    private var adapter: SpendingSectionsAdapter? = null
    private var sectionList: MutableList<SpendingSection>? = null
    private var progressView: ProgressBar? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.spendingsections_frag, container, false)

        rvSpendingSections = root.findViewById(R.id.rv_spendingsections)
        progressView = root.findViewById(R.id.spending_section_progress)

        sectionList = ArrayList()
        context?.let { adapter = SpendingSectionsAdapter(it, sectionList, presenter) }

        val mLayoutManager = GridLayoutManager(context, 1)
        rvSpendingSections!!.layoutManager = mLayoutManager
        //        rvSpendingSections.addItemDecoration(new GridSpacingItemDecoration(2, dpToPx(10), true));
        rvSpendingSections!!.itemAnimator = DefaultItemAnimator()
        rvSpendingSections!!.adapter = adapter

        presenter.requestSpendingSections()

        fabAddSpendingSection = root.findViewById(R.id.fab_spendingsecions_addspendingsection)
        fabAddSpendingSection!!.setOnClickListener {
            val intent = Intent(context, AddEditSpendingSectionActivity::class.java)
            startActivityForResult(intent, 0)
        }
        rvSpendingSections!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                if (dy > 0)
                    fabAddSpendingSection!!.hide()
                else if (dy < 0)
                    fabAddSpendingSection!!.show()
            }
        })

        // show available data quickly
        if (dataStorage.spendingSections != null) {
            dataStorage.spendingSections?.let { showSpendingSections(it.items) }
//            showSpendingSections(dataStorage.spendingSections)
//            showSpendingSections(dataStorage.spendingSections ?: null)
        }
        return root
    }

    override fun showSpendingSections(spendingSections: List<SpendingSection>) {
        adapter?.updateList(spendingSections)
    }

    override fun showErrorMessage(msg: String) {
        println("showErrorMessage! $msg")
        Snackbar.make(activity!!.findViewById(android.R.id.content), msg, Snackbar.LENGTH_LONG).show()
    }

    override fun showSpinner() {

    }

    override fun hideSpinner() {
        val animTime = resources.getInteger(android.R.integer.config_mediumAnimTime)
        showProgress(false, null, progressView, animTime)
    }

    override fun showUpdateSuccess() {
        Snackbar.make(activity!!.findViewById(android.R.id.content), R.string.spending_section_edit_success, Snackbar.LENGTH_LONG).show()
        // TODO: 05.06.2018 show mini spinner here
        presenter.requestSpendingSections()
    }

    override fun showDeleteSuccess() {
        Snackbar.make(activity!!.findViewById(android.R.id.content), R.string.spending_section_delete_success, Snackbar.LENGTH_LONG).show()
        // TODO: 05.06.2018 show mini spinner here
        presenter.requestSpendingSections()
    }

    override fun onResume() {
        super.onResume()
        presenter.takeView(this)
        //        presenter.requestSpendingSections();
        //        System.out.println("SpendingSectionsFragment onResume");
    }

    override fun onDestroy() {
        presenter.dropView()
        super.onDestroy()
    }
}
